Application/EdFi.Ods.Api/Caching/PersonUniqueIdToUsiCache.cs

```diff
@@ -290,7 +316,7 @@ private async Task InitializePersonTypeValueMapsAsync(IdentityValueMaps entry, s
                 string cacheKey = GetPersonTypeIdentifiersCacheKey(personType, context);
 
                 //Now that it's loaded extend the cache expiration.
-                _cacheProvider.Insert(cacheKey, entry, DateTime.MaxValue, TimeSpan.FromHours(4));
+                _cacheProvider.Insert(cacheKey, entry, GetAbsoluteExpiration(), _slidingExpiration);
```

Comment:
The original change that spurred all these cache issues is the switch to use ConcurrentDictionaryCacheProvider, which doesn't have any sort of cache expiration.  In this case, the parameters you're updating here are completely ignored in that implementation.
```csharp
public void Insert(string key, object value, DateTime absoluteExpiration, TimeSpan slidingExpiration)
{
    _cacheDictionary[key] = value;
}
```
The underlying cache provider would need to be updated to actually use these parameters.

Reply:
I looked at doing this, but it no longer fit into what I considered to be a simple approach, which is what we were shooting for here.

Adding a behavior that clears the entire `ConcurrentDictionary` after a particular period of time seems a bit dangerous. The cache providers are necessarily registered with the _singleton_ lifestyle so that the same instance is made available everywhere it is injected. To add a behavior that periodically clears the entire cache (for potentially many use cases, depending on the actual IoC configuration) seems like an unexpected behavior that could inadvertently cause someone trouble. As a more sensible alternative, implementing the expiration features so that it supports expiration of each cache entry certainly very quickly goes beyond the "simple" approach as well.

Another option, would be to create a copy (or override) of the `ConcurrentDictionaryCacheProvider` named `ExpiringConcurrentDictionaryCacheProvider`, and implement a simple approach that clears all the cache entries periodically. Then we'd modify the `DescriptorsCache` registration to use that version explicitly. I think it would also make sense to register it with a _transient_ lifestyle.

Another alternative would be to re-engineer the Descriptors cache to work more like the Person cache does -- to use the `AspNetCacheProvider` for a main entry for each ODS context, and then use `ConcurrentDictionary` properties on the cache entry for managing the Descriptor values. But again, this seems likely out of scope for the ticket.

A final alternative is to just not do anything right now and to put these two tickets back into the backlog and revisit after the release.

@vimayya Thoughts as to which direction you'd like this to go?

Reply:
@gmcelhanon I think ExpiringConcurrentDictionaryCacheProvider makes sense. Currently in the ConcurrentDictionaryCacheProvider's implementation of Insert method, the expiry isn't taking place which is what I wanted to address in the release. If we redesign the way we keep cache up to date later, switching back to ConcurrentDictionaryCacheProvider is easy enough.